Conversationaddition

iLevyTate opened the pull request:
Key Changes
Logging Improvements:
Updated logging configuration to include timestamps, logger names, and log levels.
Added file logging to 'app.log' for better traceability.
Context Handling:
Modified the CustomCrew class to utilize the last 5 exchanges from conversation history as context for various tasks.
Updated agent initialization to include context in their backstories.
Error Management:
Enhanced error handling in the _run_crew function.
Improved logging for different types of errors, including rate limit errors.
User Interaction Flow:
Implemented a more robust main loop that handles conversation history and follow-up questions.
Added functionality to classify user input as either a PFC task or a general task.
Code Structure:
Reorganized imports and added new dependencies as needed.
Updated class and function signatures to accommodate new features.

sanders41 commented on scan/main.py:
```diff
@@ -22,28 +27,37 @@
+client = OpenAI(api_key=openai_api_key)
+
+# Define a prompt template for general queries
+general_prompt_template = PromptTemplate(
```
Make this upper case to signify it is is a constant

```suggestion
GENERAL_PROMPT_TEMPLATE = PromptTemplate(
```

sanders41 commented on scan/main.py:
```diff
@@ -85,33 +96,97 @@ def _run_crew(crew):
+            custom_crew = CustomCrew(topic=user_input)
+            response = custom_crew.run(conversation_history)
+        else:
+            prompt = general_prompt_template.format(query=user_input)
```
```suggestion
            prompt = GENERAL_PROMPT_TEMPLATE.format(query=user_input)
```

sanders41 commented on scan/main.py:
```diff
@@ -22,28 +27,37 @@
+        self.agents = PFCAgents(llm=self.llm.llm, topic=self.topic)
+        self.tasks = PFCTasks(self.agents)
+
+    def run(self, conversation_history):
+        context = "\n".join(conversation_history[-5:])  # Use last 5 exchanges for context
+        
+        decision_task = self.tasks.complex_decision_making_task(self.topic, context)
+        emotional_task = self.tasks.emotional_risk_assessment_task(self.topic, context)
+        reward_task = self.tasks.reward_evaluation_task(self.topic, context)
+        conflict_task = self.tasks.conflict_resolution_task(self.topic, context)
+        social_task = self.tasks.social_cognition_task(self.topic, context)
```
I think I would make `conversation_history` part of the class rather than store it outside and passing it in since `CustomCrew` is the only place it is used.

```
class CustomCrew:
    def __init__(self, topic):
        self.topic = topic
        self.llm = OpenAIWrapper(api_key=openai_api_key)
        self.agents = PFCAgents(llm=self.llm.llm, topic=self.topic)
        self.tasks = PFCTasks(self.agents)
        self.conversation_history = []

    def run(self, conversation_history):
        self.conversation_history.append(conversation_history)
        context = "\n".join(self.conversation_history[-5:])  # Use last 5 exchanges for context

        decision_task = self.tasks.complex_decision_making_task(self.topic, context)
        emotional_task = self.tasks.emotional_risk_assessment_task(self.topic, context)
        reward_task = self.tasks.reward_evaluation_task(self.topic, context)
        conflict_task = self.tasks.conflict_resolution_task(self.topic, context)
        social_task = self.tasks.social_cognition_task(self.topic, context)
```

sanders41 commented on scan/main.py:
```diff
@@ -85,33 +96,97 @@ def _run_crew(crew):
+            logger.error("Received an unexpected response format from OpenAI API.")
+            return "General task"
+    except Exception as e:
+        logger.error(f"Error in classifying input: {str(e)}")
```
f-strings can have unexpected results

```suggestion
        logger.error("Error in classifying input: %s", e)

sanders41 commented on scan/main.py:
```diff
@@ -85,33 +96,97 @@ def _run_crew(crew):
+        logger.error(f"Whoops! We hit a rate limit: {e}")
+        raise
+    except Exception as e:
+        logger.error(f"Oh no! Something unexpected happened: {e}")
```
```suggestion
        logger.error("Oh no! Something unexpected happened: %s", e)
```

sanders41 commented on scan/main.py:
```diff
@@ -85,33 +96,97 @@ def _run_crew(crew):
-            raise RuntimeError(
-                "Insufficient quota. Please check your OpenAI plan and billing details."
-            ) from e
+        logger.error(f"Whoops! We hit a rate limit: {e}")
```
```suggestion
        logger.error("Whoops! We hit a rate limit: %s", e)
```

sanders41 commented on scan/main.py:
```diff
@@ -85,33 +96,97 @@ def _run_crew(crew):
+            logger.error("Hmm, looks like we didn't get a response. Let's try that again!")
+            return "I'm sorry, I couldn't generate a response. Could you please rephrase your question?"
+    except Exception as e:
+        logger.exception(f"Whoops! Something unexpected happened: {e}")
```
```suggestion
        logger.exception(f"Whoops! Something unexpected happened: %s", e)
```

sanders41 commented on scan/main.py:
```diff
@@ -85,33 +96,97 @@ def _run_crew(crew):
+                temperature=0.7
+            ).choices[0].message.content.strip()
+        if response is not None:
+            conversation_history.append(f"Bot: {response}")
```
What is this append for? It doesn't look like it ever gets used.

sanders41 commented on scan/main.py:
```diff
@@ -85,33 +96,97 @@ def _run_crew(crew):
+                    break
+            
+        except Exception as e:
+            logger.exception(f"Oops! We hit an unexpected bump: {e}")
```
```suggestion
            logger.exception(f"Oops! We hit an unexpected bump: %s", e)
```

sanders41 commented on scan/main.py:
```diff
@@ -85,33 +96,97 @@ def _run_crew(crew):
-        continue_analysis = False
-
-    print("Thank you for using the SCAN System. Have a great day!")
+    conversation_history = []
```
See other comment about adding `conversation_history` to the class.

sanders41 commented on scan/scan_agents.py:
```diff
@@ -7,27 +8,25 @@ def __init__(self, llm, topic):
         self.topic = topic
 
-    def dlpfc_agent(self):
+    def dlpfc_agent(self, context=""):
```
What if context doesn't get set when this is called? It will result in an empty string which I assume isn't what you want? An empty string is going to make debugging harder if something goes wrong. I am assuming that instead you either want to force sending a value here or set it to `None` instead of an empty string?

sanders41 commented on scan/scan_agents.py:
```diff
@@ -36,27 +35,25 @@ def dlpfc_agent(self):
         )
 
-    def vmpfc_agent(self):
+    def vmpfc_agent(self, context=""):
```
Same comment about `context=""`

sanders41 commented on scan/scan_agents.py:
```diff
@@ -65,27 +62,25 @@ def vmpfc_agent(self):
         )
 
-    def ofc_agent(self):
+    def ofc_agent(self, context=""):
```
Same comment about `context=""`

sanders41 commented on scan/scan_agents.py:
```diff
@@ -94,27 +89,25 @@ def ofc_agent(self):
         )
 
-    def acc_agent(self):
+    def acc_agent(self, context=""):
```
Same comment about `context=""`

sanders41 commented on scan/scan_agents.py:
```diff
@@ -123,27 +116,25 @@ def acc_agent(self):
         )
 
-    def mpfc_agent(self):
+    def mpfc_agent(self, context=""):
```
Same comment about `context=""`

sanders41 commented on scan/scan_tasks.py:
```diff
@@ -1,88 +1,89 @@
         self.agents = agents
 
-    def complex_decision_making_task(self, topic):
+    def complex_decision_making_task(self, topic, context=""):
```
Same comment about `context=""`

sanders41 commented on scan/scan_tasks.py:
```diff
@@ -1,88 +1,89 @@
             description=description,
             expected_output="A comprehensive analytical report with strategic recommendations.",
-            agent=self.agents.dlpfc_agent(),
+            agent=self.agents.dlpfc_agent(context),
```
Same comment about `context=""`

sanders41 commented on scan/scan_tasks.py:
```diff
@@ -1,88 +1,89 @@
         )
 
-    def emotional_risk_assessment_task(self, topic):
+    def emotional_risk_assessment_task(self, topic, context=""):
```
Same comment about `context=""`

iLevyTate commented:
### main.py
**Changes: Conversation History Management**
- Moved `conversation_history` into the `CustomCrew` class.
- Updated the `run` method to use `self.conversation_history`.
- Modified `handle_input` to accept a `CustomCrew` instance.
- Removed unnecessary `append` operation in `handle_input`.
- Adjusted `main` to create and reuse a single `CustomCrew` instance.

### scan_agents.py
**Changes: Context Integration**
- Added `context` parameter (default `None`) to all agent methods.
- Incorporated `context` into agent backstories.
- Updated agent calls to pass `context`.

### scan_tasks.py
**Changes: Context Handling**
- Added `context` parameter (default `None`) to all task methods.
- Incorporated `context` into task descriptions.
- Updated agent calls to pass `context`.
- Maintained clear "Expected Output" sections.

sanders41 reviewed (approved):
lgtm
